### DISCUSSION TITLE
VM migration time

### DISCUSSION THREAD

**Discussion by kajmera31** [manoelcampos/cloudsim-plus#289] [Q&A] [answered]
How much time it will take to migrate a VM? 

The default time interval is 5 minutes (300 Sec). However I read the paper "Minarolli, D., Mazrekaj, A. and Freisleben, B., Tackling uncertainty in long-term predictions for host overload and underload detection in cloud computing. Journal of Cloud Computing, 6(1), p.4., 2017", they mentioned that on average it will take approximately 4-time interval to migrate VMs. Based on repeated simulation they come to this value.
As per my understanding, at the beginning of each time interval, all migrations are performed.
Thanks for your kind consideration.

**Comment by manoelcampos** [comment, marked as answer]
I don't know the paper, but you may be confusing things. Probably, the paper is using PlanetLab trace files containing CPU utilization data for a given VM. Data inside those files were collected in 5 minutes intervals. This way, if you run a simulation where a VM uses some of these files for its CPU utilization model, by defining the Datacenter schedulingInterval as 300 seconds, at every 5 minutes the CPU utilization for that VM will be equal to the next line inside that file.
If your schedulingInterval is different from 300, a mean between the two closest rows inside the file is computed to provide the current VM CPU utilization percentage.
This way, I don't know what you mean by one-time interval.

The time intervals to migrate a VM is totally dependent on multiple factors such as the VM's ram size (capacity), available BW for VM migration (check Datacenter.DEF_BW_PERCENT_FOR_MIGRATION with is 50%) and the Datacenter schedulingInterval.

Consider you have:

- a VM with 20GB of RAM capacity;
- total BW of 2GB/s (using default config as presented above, that means only 1GB/s will be available for VM migration);
- a Datacenter schedulingInterval of 2.5 seconds

the VM migration will take 20 seconds. In this scenario, it will pass 8 intervals of 2.5 seconds until the VM migration finishes.

Therefore, the migration time and number of intervals to finish migration is completely variable. But the later one is just relevant for simulation purposes. In actual cloud infrastructures, there is not such a thing as "time intervals" for VM migrations. The migration occurs as a continuous event that starts and finishes at a given specific time. At the end, we just have a time spent for migration, not multiple fixed time intervals.

**Reply by kajmera31** [reply]
Thank you so much Manoel for a detailed explanation, it resolves my query.

**Reply by manoelcampos** [reply]
You're welcome. Please mark the discussion as answered and give the project a star 😉
